tools/c7n_mailer/c7n_mailer/gcp_mailer/gcp_queue_processor.py

```diff
@@ -0,0 +1,118 @@
+# Copyright The Cloud Custodian Authors.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Google Queue Message Processing
+==============================
+
+"""
+import base64
+import json
+import zlib
+
+import six
+from c7n_mailer.email_delivery import EmailDelivery
+
+try:
+    from c7n_gcp.client import Session
+except ImportError:
+    raise Exception("Using GCP Pub/Sub with c7n_mailer requires package c7n_gcp to be installed.")
+
+MAX_MESSAGES = 1000
+
+
+class MailerGcpQueueProcessor(object):
+    def __init__(self, config, logger, session=None):
+        self.config = config
+        self.logger = logger
+        self.subscription = self.config["queue_url"]
+        self.session = session or Session()
+        self.client = self.session.client("pubsub", "v1", "projects.subscriptions")
+
+    def run(self):
+        self.logger.info("Downloading messages from the GCP PubSub Subscription.")
+
+        # Get first set of messages to process
+        messages = self.receive_messages()
+
+        while len(messages) > 0:
+            # Discard_date is the timestamp of the last published message in the messages list
+            # and will be the date we need to seek to when we ack_messages
+            discard_date = messages["receivedMessages"][-1]["message"]["publishTime"]
+
+            # Process received messages
+            for message in messages["receivedMessages"]:
+                self.process_message(message)
+
+            # Acknowledge and purge processed messages then get next set of messages
+            self.ack_messages(discard_date)
+            messages = self.receive_messages()
+
+        self.logger.info("No messages left in the gcp topic subscription, now exiting c7n_mailer.")
+
+    # This function, when processing gcp pubsub messages, will deliver messages over email.
+    # Also support for Datadog and Slack
+    def process_message(self, encoded_gcp_pubsub_message):
+        pubsub_message = self.unpack_to_dict(encoded_gcp_pubsub_message["message"]["data"])
+        # Process email first
```

Comment:
This method is very similar to `process_sqs_message`.  Would it be better to have an abstract class for `MailerSqsQueueProcessor`, `MailerGcpQueueProcessor` and `MailerAzureQueueProcessor`? So that most of the logic of `process_sqs_message` can be reused?

Different kinds of queues are carriers to carries the same payload (notification), so we should have sub classes to unwrap the payload and have the same class/method to process the payload.